### PR TITLE
fetch-improvements

### LINKED ISSUE #66: Improvements for "codemon fetch"
- When inputs and outputs are written in the `.in` and `.op` files, there's an additional newline at the top of each file, which can cause problems in diffing between outputs during `codemon listen`, therefore, we need to avoid adding that newline either during write or be remove post write.
- Improve all verbose messages and use consistent coloring, currently, some of them are green and some are yellow, in a very inconsistent way. Maybe stick to white or some other bright colour which will be good for a basic terminal.
- Maybe get rid of `codemon fetch <contestName>` ? This needs discussion. Currently there are 3 ways to leverage the fetching functionality: `codemon fetch` inside contest directory, `codemon init -f <contestdir>`, `codemon fetch <contestName>`. The last one creates a directory with given contest name and contains only inputs and outputs and no code files. I believe we have already served that purpose with `codemon init -f <contestDir>`, therefore we can get rid of this command.

### PR DESCRIPTION
**Pull request by Enigmage**
Closes #66 

#### Required Changes

- Removes `codemon fetch <contest_name>` command as that is redundant to the functionality of codemon, especially after the addition of  `-f /--fetch` flag. Also added error checks so that `codemon fetch` is entered as a single command.
-  Every sample test case string after scraping from codeforces  begins with a  `\n` character. So to solve the problem of an extra newline at the beginning of inputs and outputs, all that was to be done was to write the `i` and `o` strings to input and output files from the 1st index instead of the 0th. Example.
```py
f.write(i[1:])
```
- As suggested changed the coloring of all verbose messages to white, except error messages which will be red.

#### Additional Changes

There were a few inconsistencies in the functionality of codemon and `CodemonFetch` module that I've fixed.

- It was the intended behaviour that the user can enter any string as the contest name as long  it contained the correct contest number in it for example `contest_1455` or `round_1455`. Currently if the user enters this as the contest name during init they will get the correct filenames for round 1455 but if they were to run `codemon fetch`  inside this directory instead of the sample test cases being written down into their correct place, an extra folder would be created with the contest number as its name, that contained only the inputs and outputs. This is because unlike `get_filename` function in CodemonMeta.py, `fetch_tests` function did not clearly differentiate between the contest number and contest name while scraping. This created the issue. So with this update I've ensured that clear distinction is made between the contest number and contest name. So the user can name their contest directory like `round_1455` etc without facing any issue.

- Since init now takes care of making the proper directory structure the `make_structure` function in `CodemonFetch` module was redundant. So I changed it to `check_structure` which only checks for the presence of correct directory structure while scraping, if the correct structure is not present it will generate an error. Which should be the correct behaviour. So if the user now accidentally runs `codemon fetch` in some other random directory that happened to have a valid codeforces contest number in its name, codemon fetch will now error out instead of creating a question folder with inputs and outputs and crowding up the users directory. This can also act as a very good error check as to whether the user is running `codemon fetch` from inside contest directory or not.

- Added better error checks to `fetch_tests` function, so if fetching of sample test cases fails user will know exactly why.
- There were few unused imports, redundant variables and code that I've removed to make the code more concise.

### CODE REVIEW

**Comment by ankushbhardwxj** [comment]
@Enigmage - Sorry for delay in review. Could you also work on improving the documentation (how to use commands, naming system for directory, a GIF demonstrating codemon, etc) on our README and make a new PR for that too ?

**Comment by Enigmage** [comment]
@ankingcodes Yeah sure I can work on that. Can you just give me a rough idea of what all you'd want to be included in the README and the Gif so that I can start working accordingly ?

**Comment by ankushbhardwxj** [comment]
Take each of the commands for example and desribe each of them in detail. For example, for `codemon init -f round_1342`, you can explain what the command does, how many ways the command can be written, for example it can be written as `codemon init --fetch 1342 ` also. Then you can go on desribing the proper directory nomenclature etc. 

I'd be much happy if you can generate docs from the code itself, maybe using something like [pydocs](https://docs.python.org/3/library/pydoc.html). Do it whichever way suits you better.

**Comment by Enigmage** [comment]
@ankingcodes Hey I forgot to ask, are you expecting the documentation changes in this PR itself ? Because I was was hoping you could review and if its all fine merge this PR first since it has some important fixes for `codemon fetch`. And then I will be making a PR for adding documentation separately (already working on it on a different branch). Is this okay?

UPDATE: I also found some code that could give error on windows. This for example
```py
os.getcwd().split('/')[-1]
```
This might error out since windows doesn't have forward slashes.  So I have replaced it with an OS agnostic way of extracting current directory name:

```py
os.path.basename(os.getcwd())
```

File: codemon/CodemonInit.py [hunk 1]
```diff
@@ -12,14 +12,14 @@ def write_to_file(filename, text, contestName=None):
...
+    print(f'Creating directories for contest {contestName}...')
     os.makedirs(os.path.join(os.getcwd(), contestName))
     for f in fileNames:
       os.makedirs(os.path.join(os.getcwd(), contestName, f))
   except OSError:
-    print("Failed! This directory cannot be created.")
+    print(colored.red("Failed! This directory cannot be created."))
   else:
-    print(colored.yellow('Directories are made.'))
+    print('Directories are made.')
```
Comment (ankushbhardwxj) [inline codemon/CodemonInit.py]: Can we have only one output "Created directory {contestName}" for creation of directories ? And "Failed to create directory {contestName} !" ?

File: codemon/CodemonInit.py [hunk 2]
```diff
@@ -44,4 +44,4 @@ def init_single_file(filename, init_flags):
   print("Making single file")
```
Comment (ankushbhardwxj) [inline codemon/CodemonInit.py]: Please get rid of this line.

File: codemon/CodemonInit.py [hunk 3]
```diff
@@ -44,4 +44,4 @@ def init_single_file(filename, init_flags):
   print("Making single file")
   with open(full_filename, 'w+') as f:
     f.write(use_template)
-  print(colored.yellow(f"Created {filename}.{ext}"))
+  print(f"Created {filename}.{ext}")
```
Comment (ankushbhardwxj) [inline codemon/CodemonInit.py]: Rewrite this as : `Created file {filename}.{ext}`

File: codemon/CodemonHelp.py [hunk 4]
```diff
@@ -17,8 +17,7 @@ def showHelp():
   print("codemon - - - - - - - - - - - - - - -  displays this message")
   print("codemon --help - - - - - - - - - - - - shows help")
   print("codemon init <contestName> - - - - - - initialises a contest")
-  print("codemon fetch <contestName> - - - - fetches all the sample test cases for the contest name provided")
-  print("codemon fetch  - - - - - - - - - - - extracts contest name from contest directory name and fetches all sample test cases ")
+  print("codemon fetch  - - - - - - - - - - - fetch all sample test cases for the current contest directory")
```
Comment (ankushbhardwxj) [inline codemon/CodemonHelp.py]: Please rewrite this as : `fetches testcases from CodeForces`.

File: codemon/CodemonFetch.py [hunk 5]
```diff
@@ -1,58 +1,53 @@
...
-          with open(os.path.join(basedir, f'{file_name}' , f'{file_name}.in'), 'a') as f:
-            f.write(i)
-
-        # Add outputs to .op files
-        for t in test.findAll("div", attrs={"class":"output"}):
-          o = t.pre.text
-          with open(os.path.join(basedir, f'{file_name}' , f'{file_name}.op'), 'a') as f:
-            f.write(o)
-      print(colored.yellow("Inputs and Outputs added"))
+          flag = False
```
Comment (ankushbhardwxj) [inline codemon/CodemonFetch.py]: change the flag variable to some proper name, for example, `correctDirStruct` ?

**Review by ankushbhardwxj — CHANGES_REQUESTED** [review]

**Comment by ankushbhardwxj** [comment]
@Enigmage - few minor changes required here.

File: codemon/CodemonInit.py (hunk 1 above)
Comment (Enigmage) [reply codemon/CodemonInit.py]: You are right, it is cleaner this way.

**Comment by Enigmage** [comment]
@ankingcodes made the requested changes. Also added info about the `-f` flag for fetch in the Help section which was missing, see if you like it. Couldn't include `--fetch` as the message was getting long.

**Review by ankushbhardwxj — APPROVED** [review]
Looks good ! Thanks @Enigmage

**Pull request merged**